syntax = "proto3";

package flower_sdk;

service FlowerServiceSDK {
    rpc Join (stream ClientMessage) returns (stream ServerMessage) {}
}

message Parameters {
    repeated bytes parameters = 1;
    string tensor_type = 2;
}

enum Reason {
    UNKOWN = 0;
    RECONNECT = 1;
    POWER_DISCONNECTED = 2;
    WIFI_UNAVAILABLE = 3;
    ACK = 4;
}

message ServerMessage {

    message Reconnect { int64 seconds = 1; }

    message GetParameters {}

    message FitIns {
        Parameters parameters = 1;
        map<string, Scalar> config = 2;
    }

    message EvaluateIns {
        Parameters parameters = 1;
        map<string, Scalar> config = 2;
    }

    message PropertiesIns { map<string, Scalar> config = 1; }

    oneof msg {
        Reconnect reconnect = 1;
        GetParameters get_parameters = 2;
        FitIns fit_ins = 3;
        EvaluateIns evaluate_ins = 4;
        PropertiesIns properties_ins = 5;
    }
}

message ClientMessage {

    message Disconnect { Reason reason = 1; }

    message ParametersRes { Parameters parameters = 1; }

    message FitRes {
        Parameters parameters = 1;
        int64 num_examples = 2;
        map<string, Scalar> metrics = 5;
    }

    message EvaluateRes {
        int64 num_examples = 1;
        float loss = 2;
        map<string, Scalar> metrics = 4;
    }

    message PropertiesRes { map<string, Scalar> properties = 1; }

    oneof msg {
        Disconnect disconnect = 1;
        ParametersRes parameters_res = 2;
        FitRes fit_res = 3;
        EvaluateRes evaluate_res = 4;
        PropertiesRes properties_res = 5;
    }
}

message Scalar {
    oneof scalar {
        double double = 1;
        sint64 sint64 = 2;
        bool bool = 3;
        string string = 4;
        bytes bytes = 5;
    }
}
